src/testing/fuzz.rs

```diff
@@ -0,0 +1,299 @@
+pub use crate::testing::mock::NetworkData;
+use crate::{
+    nodes::NodeIndex,
+    testing::mock::{
+        configure_network, spawn_honest_member, Data, Hasher64, NetworkHook, PartialMultisignature,
+        Signature, Spawner,
+    },
+    Network, NetworkData as ND, SpawnHandle,
+};
+use codec::{Decode, Encode, IoReader};
+use futures::{
+    channel::oneshot::{self, Receiver},
+    StreamExt,
+};
+use futures_timer::Delay;
+use log::{error, info};
+use std::{
+    io::{BufRead, BufReader, Read, Result as IOResult, Write},
+    iter::once_with,
+    time::Duration,
+};
+use tokio::runtime::{Builder, Runtime};
+
+struct ClosureHook<F> {
+    wrapped: F,
+}
+
+impl<F: FnMut(&NetworkData, NodeIndex, NodeIndex) + Send> ClosureHook<F> {
+    pub(crate) fn new(wrapped: F) -> Self {
+        ClosureHook { wrapped }
+    }
+}
+
+impl<F: FnMut(&NetworkData, NodeIndex, NodeIndex) + Send> NetworkHook for ClosureHook<F> {
+    fn update_state(&mut self, data: &mut NetworkData, sender: NodeIndex, recipient: NodeIndex) {
+        (self.wrapped)(data, sender, recipient);
+    }
+}
+
+#[derive(Default)]
+pub struct NetworkDataEncoding {}
+
+impl NetworkDataEncoding {
+    pub fn encode_into<W: Write>(&self, data: &NetworkData, writer: &mut W) -> IOResult<()> {
+        writer.write_all(&data.encode()[..])
+    }
+
+    pub fn decode_from<R: Read>(
+        &self,
+        reader: &mut R,
+    ) -> core::result::Result<NetworkData, codec::Error> {
+        let mut reader = IoReader(reader);
+        <NetworkData>::decode(&mut reader)
+    }
+}
+
+struct NetworkDataIterator<R> {
+    input: R,
+    encoding: NetworkDataEncoding,
+}
+
+impl<R: Read> NetworkDataIterator<R> {
+    fn new(read: R) -> Self {
+        NetworkDataIterator {
+            input: read,
+            encoding: NetworkDataEncoding::default(),
+        }
+    }
+}
+
+impl<R: Read> Iterator for NetworkDataIterator<R> {
+    type Item = NetworkData;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.encoding.decode_from(&mut self.input) {
+            Ok(v) => Some(v),
+            Err(e) => {
+                error!("Error while decoding NetworkData: {:?}.", e);
+                None
+            }
+        }
+    }
+}
+
+struct RecorderNetwork<I: Iterator<Item = NetworkData>> {
```

Comment:
The name is a bit confusing, maybe `PlaybackNetwork`?

Reply:
changed